### ISSUE TITLE
Import declaration problem with module containing typings

### ISSUE THREAD

**Issue by anthonybriand** [dsherret/ts-simple-ast#122]
There seems to be a problem when I reference a class, an interface, a decorator from a module that is in node_modules and that contains an index.d.ts file.

I create a SourceFileStructure, and when I add the namedImports referring to a module containing an index.d.ts file, it throws this error : 
```
/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/utils/EventContainer.js:53
            finally { if (e_1) throw e_1.error; }
                               ^

Error: Debug Failure. False expression.
    at tryReuseStructureFromOldProgram (/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/node_modules/typescript/lib/typescript.js:72065:22)
    at Object.createProgram (/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/node_modules/typescript/lib/typescript.js:71808:34)
    at Program.reset (/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/compiler/tools/Program.js:31:35)
    at LanguageService.resetProgram (/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/compiler/tools/LanguageService.js:112:26)
    at LanguageService.addSourceFile (/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/compiler/tools/LanguageService.js:252:14)
    at /Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/GlobalContainer.js:21:40
    at EventContainer.fire (/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/utils/EventContainer.js:45:17)
    at /Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/factories/CompilerFactory.js:136:49
    at KeyValueCache.getOrCreate (/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/utils/KeyValueCache.js:17:20)
    at CompilerFactory.getSourceFile (/Users/acidspike/Development/Sources/SCM/External/Harmony/Orm/node_modules/ts-simple-ast/dist/factories/CompilerFactory.js:128:31)
```

Basic SourceFileStructure sample : 
```
    {
        classes: [
            {
                name: "Aze",
                isExported: true,
                decorators: [
                    {
                        name: "Entity",
                        arguments: [
                            `
                            {
                              "connectionOptions": {
                                  "connectionName": "default"
                              },
                              "name": "aze"
                          }
                            `
                        ]
                    }
                ],
                properties: [],
                methods: [],
                "extends": "BaseModel"
            }
        ],
        imports: [
            {
                namedImports: [
                    {
                        "name": "Entity"
                    },
                    {
                        "name": "BaseModel"
                    }
                ],
                moduleSpecifier: "@harmony/orm" //<-- this throw the error because the module contains an index.d.ts file
            }
        ]
    }
```

On the other hand the use of the @harmony/orm in a project (except AST) works correctly no error and typescript well understand the .d.ts

**Comment by dsherret** [comment]
I'm not sure why that errors. It seems like TypeScript is having trouble creating a new program from the old one. On line 54 in [Program.ts}(https://github.com/dsherret/ts-simple-ast/blob/774bf701afb0bedfba61965bfb26e31e609d3b63/src/compiler/tools/Program.ts#L54), probably the last argument shouldn't bother to be passed in (`this._compilerObject`). That seems like just an optimization anyway.

It would be good to know why that "debug failure" happens internally in the TypeScript compiler though.

**Comment by anthonybriand** [comment]
I located the line that throw the error, it's in the tryReuseStructureFromOldProgram
exactly this line : ts.Debug.assert(!(oldProgram.structureIsReused & (2 /* Completely */ | 1 /* SafeModules */)));

In my case the result of the statement : !(oldProgram.structureIsReused & (2 /* Completely */ | 1 /* SafeModules */)) is 1

Now why this is a problem, I do not know. But with the remove of the this._compilerObject parameter of the createProgram call, this works fine.

**Comment by dsherret** [comment]
Should be fixed in 0.87.0.

This is a bit of a band aid. If the first way doesn't succeed then it will fall back to trying without using the existing compiler object.

**Comment by anthonybriand** [comment]
I confirm, it works, thank you.
